core/drivers/crypto/hisilicon/hpre_dh.c

```diff
@@ -0,0 +1,380 @@
+{
+	struct hpre_dh_msg *msg = (struct hpre_dh_msg *)info;
+	struct hpre_sqe *sqe = (struct hpre_sqe *)bd;
+	uint16_t err = 0;
+	uint16_t done = 0;
+
+	err = HPRE_TASK_ETYPE(sqe->w0);
+	done = HPRE_TASK_DONE(sqe->w0);
+	if (done != HPRE_HW_TASK_DONE || err) {
+		EMSG("HPRE do dh fail! done=0x%x, etype=0x%x", done, err);
+		return HISI_QM_DRVCRYPT_IN_EPARA;
+	}
+
+	if (hpre_bin_to_crypto_bin(msg->out, msg->out, msg->key_bytes,
+				   msg->out_bytes)) {
+		EMSG("Fail to transfer dh_y from hpre_bin to crypto_bin");
+		return HISI_QM_DRVCRYPT_EINVAL;
+	}
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static TEE_Result hpre_dh_do_task(void *msg)
+{
+	struct hisi_qp *qp;
+	TEE_Result res = 0;
+	int32_t ret;
+
+	qp = hpre_create_qp(HISI_QM_CHANNEL_TYPE0);
+	if (!qp) {
+		EMSG("Fail to create dh qp!");
+		return TEE_ERROR_BUSY;
+	}
+
+	qp->fill_sqe = hpre_dh_fill_sqe;
+	qp->parse_sqe = hpre_dh_parse_sqe;
+	ret = hisi_qp_send(qp, msg);
+	if (ret) {
+		EMSG("Fail to send task, ret=%d!", ret);
+		res = TEE_ERROR_BAD_STATE;
+		goto done_proc;
+	}
+
+	ret = hisi_qp_recv_sync(qp, msg);
+	if (ret) {
+		EMSG("Recv task error, ret=%d!", ret);
+		res = TEE_ERROR_BAD_STATE;
+		goto done_proc;
+	}
+
+done_proc:
+	hisi_qm_release_qp(qp);
+
+	return res;
+}
+
+static size_t round_key_size_to_hw_size(size_t key_bits)
+{
+	size_t size = 0;
+
+	if (key_bits <= 768)
+		size = BITS_TO_BYTES(768);
+	else if (key_bits <= 1024)
+		size = BITS_TO_BYTES(1024);
+	else if (key_bits <= 1536)
+		size = BITS_TO_BYTES(1536);
+	else if (key_bits <= 2048)
+		size = BITS_TO_BYTES(2048);
+	else if (key_bits <= 3072)
+		size = BITS_TO_BYTES(3072);
+	else if (key_bits <= 4096)
+		size = BITS_TO_BYTES(4096);
+	else
+		EMSG("Invalid key_bits[%lu]!", key_bits);
+
+	return size;
+}
+
+static TEE_Result hpre_dh_gen_privkey(struct bignum *x, size_t key_bits)
+{
+	size_t key_bytes = BITS_TO_BYTES(key_bits);
+	uint8_t buf[HPRE_DH_MAX_KEY_BYTES] = { };
+	TEE_Result ret;
+
+	if (hw_get_random_bytes(buf, key_bytes)) {
+		EMSG("Fail to fill privkey!");
+		return TEE_ERROR_NO_DATA;
+	}
+
+	ret = crypto_bignum_bin2bn(buf, key_bytes, x);
+	memzero_explicit(buf, key_bytes);
+
+	return ret;
+}
+
+static int32_t hpre_dh_params_alloc(struct hpre_dh_msg *msg)
+{
+	uint32_t size = HPRE_DH_TOTAL_BUF_SIZE(msg->key_bytes);
+	uint8_t *data;
+
+	data = (uint8_t *)calloc(1, size);
+	if (!data) {
+		EMSG("Fail to alloc dh total buf!");
+		return HISI_QM_DRVCRYPT_ENOMEM;
+	}
+
+	msg->x_p = data;
+	msg->x_p_dma = virt_to_phys(msg->x_p);
+	if (!msg->x_p_dma) {
+		EMSG("Fail to get x_p dma addr!");
+		free(data);
+		msg->x_p = NULL;
+		return HISI_QM_DRVCRYPT_EFAULT;
+	}
+
+	msg->g = msg->x_p + (msg->key_bytes << 1);
+	msg->g_dma = msg->x_p_dma + (msg->key_bytes << 1);
+	msg->out = msg->g + msg->key_bytes;
+	msg->out_dma = msg->g_dma + msg->key_bytes;
+
+	return 0;
+}
+
+static void hpre_dh_params_free(struct hpre_dh_msg *msg)
+{
+	if (msg->x_p) {
+		memzero_explicit(msg->x_p, msg->key_bytes << 1);
```

Comment:
A job for `free_wipe()`?

Reply:
Should I use free_wipe(msg->x_p) here?

Reply:
Yes, it should get the job done, provided that the buffer size is `msg->key_bytes << 1` or close enough.

Reply:
the buffer size is 4 * msg->key_bytes << 1, but we only set x and p to zero.

Reply:
Does it matter?

Reply:
solved